Stm32: Add HPDMA/HSEM/FMC/IPCC drivers

GseoC opened the pull request:
These peripherals are aware of the hardware peripheral firewall on the STM32MP25 platform. Therefore, add a driver for each of them to manage their firewall configuration.

Should these drivers be compatible with STM32MP1x platforms or manage other peripheral features, content will be added in them.

Nodes are default disabled for now as the clock driver is not yet upstreamed.

anilreddy3 commented on core/drivers/stm32_hsem.c:
```diff
@@ -0,0 +1,283 @@
+
+	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
+	if (!cuint)
+		panic("No RIF configuration available");
```
Would it be better to print error message and return instead of panic here, as it is initial driver probe dt parsing.

GseoC replied:
I can do that. IMO it's pretty much the same because in the end, the platform will panic. I'd rather panic right now to avoid further useless code execution.
Maybe @etienne-lms can give us a hint on what to do.

etienne-lms replied:
Failing to apply a firewall an awaited configuration directive is, I think, I good reason to refuse to boot, hence panic.

anilreddy3 commented on core/drivers/stm32_hsem.c:
```diff
@@ -0,0 +1,283 @@
+
+	for (i = 0; i < hsem_d->nb_channels; i++) {
+		rif_conf = fdt32_to_cpu(cuint[i]);
+
```
Suggestion:

```
stm32_rif_parse_cfg(fdt32_to_cpu(cuint[i]), &hsem_d->conf_data,
				    HSEM_NB_MAX_CID_SUPPORTED,
				    HSEM_RIF_RESOURCES);
```

and to remove extra empty lines.

GseoC commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,488 @@
+			fmc_d->cclken = true;
+
+		if (!fmc_d->cclken)
+			break;
```
Yes, it should be continue

GseoC commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,488 @@
+
+		/* Check if a controller is secure */
+		if (is_fmc_controller_secure(i))
+			return TEE_ERROR_BAD_STATE;
```
I don't understand, it's fine as is IMO

etienne-lms replied:
Don't you need to balance `clk_enable()` instruction from line 305?
Ditto at line 328, a `break` or `goto end` instruction would do the job.

GseoC commented:
Updated with comments addressed

etienne-lms commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,488 @@
+			continue;
+
+		if (fdt_read_uint32(fdt, ctrl_node, "reg", &bank) < 0)
+			return TEE_ERROR_BAD_PARAMETERS;
```
Ok, fine as is. But it's a better habit to free memory from a place related to where it was allocated.
No big deal here, but usually easier to track heap leakage.

etienne-lms commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,488 @@
+		unsigned long hclkp = FMC_NSEC_PER_SEC / (hclk / 1000);
+		unsigned long timing = DIV_ROUND_UP(fmc_d->clk_period_ns * 1000,
+						    hclkp);
+		uint32_t clk_div = 1 << _FMC_CFGR_CLKDIV_SHIFT;
```
should use `SHIFT_U32()` for 32bit left shift operations.
Ditto at line 383.

etienne-lms commented on core/drivers/stm32_hpdma.c:
```diff
@@ -0,0 +1,349 @@
+	int lenp = 0;
+	const fdt32_t *cuint = NULL;
+	struct dt_node_info info = {};
+	struct io_pa_va addr = {};
```
s/{}/{ }/

etienne-lms commented on core/drivers/stm32_hpdma.c:
```diff
@@ -0,0 +1,349 @@
+			if (res) {
+				EMSG("Couldn't release semaphore for res%u", i);
+				return TEE_ERROR_ACCESS_DENIED;
+			}
```
here, if `res == TEE_SUCCESS`, we still failed to aquire the semaphore (test at line 150). Shouldn't this also return a error code?

GseoC replied:
Mmm i don't think so, the goal here is to release any semaphore that could have been taken by the CID1 if it's not in the authorized list anymore.

etienne-lms replied:
ok

etienne-lms commented on core/drivers/stm32_hpdma.c:
```diff
@@ -0,0 +1,349 @@
+
+static void stm32_hpdma_pm_resume(struct hpdma_pdata *hpdma)
+{
+	apply_rif_config(hpdma, true);
```
panic on failure (e.g. failing to acquire a semaphore): there is alternate way on a PM resume operation.

GseoC replied:
I guess you meant "no alternate"

etienne-lms replied:
Right, sorry, "no alternate way".

GseoC commented:
Updated with comments addressed

etienne-lms commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,491 @@
+		panic("Couldn't acquire controller 0 semaphore");
+
+	if (fmc_d->pinctrl_d)
+		pinctrl_apply_state(fmc_d->pinctrl_d);
```
Sorry, I missed this one: `pinctrl_apply_state()` returns a `TEE_Result` code, it should be checked.
Ditto at line 420.

etienne-lms commented on core/drivers/stm32_hpdma.c:
```diff
@@ -0,0 +1,350 @@
+
+	register_pm_core_service_cb(stm32_hpdma_pm, hpdma_d, "stm32-hpdma");
+
+	return res;
```
nitpicking: `return TEE_SUCCESS`

etienne-lms commented on core/drivers/stm32_hsem.c:
```diff
@@ -0,0 +1,279 @@
+
+	hsem_d->rif_proc_conf = calloc(HSEM_NB_PROC, sizeof(uint32_t));
+	if (!hsem_d->rif_proc_conf)
+		panic("Missing memory capacity for HSEM RIF proc configuration");
```
nitpicking: `assert(hsem_d->rif_proc_conf)`  should be enough (ditto at lines 209/211), as for other memory allocation instructions.

GseoC replied:
I'd rather return TEE_ERROR_OUT_OF_MEMORY

GseoC replied:
I'm propagating this to drivers in this P-R

GseoC replied:
mmmm actually I'd rather use the assertion

etienne-lms reviewed (commented):
Few remaining comments. I'm fine you squash all fixes.

GseoC commented:
Pushed with comments addressed and added `__maybe_unused` to `check_fmc_rif_conf()`

etienne-lms commented on core/drivers/stm32_hpdma.c:
```diff
@@ -0,0 +1,349 @@
+			EMSG("Couldn't acquire semaphore for channel %u", i);
+			return res;
+		}
+	}
```
nitpicking: better with an empty line below.

etienne-lms commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,490 @@
+
+	for (i = 0; i < FMC_RIF_CONTROLLERS; i++) {
+		if (!(BIT(i) & fmc_d->conf_data.access_mask[0]))
+			continue;
```
nitpicking: better with an empty line below.

etienne-lms commented on core/drivers/stm32_fmc.c:
```diff
@@ -0,0 +1,490 @@
+			continue;
+
+		/* If not TDCID, we want to acquire semaphores assigned to us */
+
```
Could you remove this empty line?

etienne-lms reviewed (commented):
For commit "drivers: add FMC driver with RIF support":
- 1 return value to check + few minor comments.

For commit "drivers: add HPDMA driver with RIF support":
- 1 minor comment.

For commit "drivers: add RIF support driver for HSEM":
- in commit message, sentence "In order to apply CID filtering over a group of semaphore." seems to lack a verb. I would replace "In order to" with "Used to", but my English is quite weak so this is not a strong opinion.

With these addressed, you cna apply my 
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` tag.

etienne-lms commented on core/drivers/stm32_hsem.c:
```diff
@@ -0,0 +1,274 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2022, STMicroelectronics
```
I missed this one: `2022-2024`

GseoC commented:
Rebased with comments addressed and tag applied.

I also changed the license of the drivers from `BSD-3-Clause` to `BSD-2-Clause` and was legitimate to do so.

Thanks
